### DISCUSSION TITLE
Requirments for MonkeyPoll

### DISCUSSION THREAD

**Discussion by yehandesilva** [yehandesilva/monkeyPoll#16] [General]
Add any ideas for features and requirements here. We can discuss and finalize them into a solid set of requirements that we can distribute among our milestones.

**Comment by pathumd** [comment]
- The user is not required to create an account to submit a survey response
- The user shall access a survey by navigating to the survey's URL
- In order for the backend to uniquely identify a user's survey response, the user shall provide their email address (along with their first name and last name). The assumption here is that a user's email can only submit one response per survey

**Reply by yehandesilva** [reply]
In that case, we need to ensure survey responses contains an email field which must be a valid email and can't be null.

**Reply by AminZeina** [reply]
It sounds like we'll need a "survey response" entity which has fields for email, first name, last name, the survey ID, and a list of question responses.

Which means that going further, we would need some "question response" entity for each question type. So a survey has a list of "questions" (each of which will actually be one of three question types), and then a response has a list of "question responses" (which again would actually be one of three question type responses). Does that make sense?

**Reply by pathumd** [reply]
Yes I think that makes sense. In terms of modelling the 3 types of 'questions', and 3 types of 'question responses', we could create a generic 'Question' and 'QuestionResponse' class (which contains common attributes) to which other types can inherit from. For example, 'TextBasedQuestion', 'NumericalRangeQuestion' and 'SelectionQuestion' could extend 'Question', etc. 

For reference: The following articles show an example of creating entities with inheritance: 
https://www.baeldung.com/hibernate-inheritance
https://www.javatpoint.com/jpa-inheritance-overview

**Reply by AminZeina** [reply]
That sounds good, and the references look like exactly what we need, nice!

**Reply by yehandesilva** [reply]
Thanks for the documentation @pathumd, helped me when creating the UML diagrams for our alpha release milestone. I can't use the @MappedSuperclass annotation since our Question class will have a relationship with Response class, which the annotation can't handle. Will need to use one of the other methods menitioned in the first link.

**Comment by yehandesilva** [comment]
According to the project description, these are the requirements it lists that we absolutely need:
1) User can create a survey.
2) Other users can access and fill out the survey.
3) Survey can be closed by the creator at their discretion preventing any new responses.
4) Upon closing a survey, the creator gets a detailed view of all their responses.

**Comment by yehandesilva** [comment]
Types of questions:
| Type of question         | Result View  |
|--------------------------|--------------|
| Open-ended text          | Listed as is |
| Number within a range    | Histogram    |
| Multiple choice question | Pie chart    |

**Reply by AminZeina** [reply]
Each of these questions should be an entity with some "question" parent entity/superclass. A survey should have 1...* questions, and each question should be one of the listed **Type of question** above.

As for displaying results, there should be an endpoint that the user can use to view a survey's results; we'll need a controller to get the survey's results from the DB and display the results for each question based on the **Result View** above.

**Comment by AminZeina** [comment]
These requirements for the core actions that the user should be able to perform:
- Users shall be able to create a create a new account
- Users shall be able to login to their account
- Logged in users shall be able to create new surveys
- Logged in users shall be able to close surveys that they created at any time
- When creating a survey, the user should be able to add multiple questions of different types
- Users can fill in surveys without logging in
